File: backend/app/routers/reports.py
import io

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import require_roles
from app.models.employee import Employee
from app.schemas.dashboard import (
    BookingHeatmap,
    DepartmentSummaryRow,
    MaintenanceFrequencyRow,
    RetirementForecastRow,
    UtilizationRow,
)
from app.services import report_service

router = APIRouter(prefix="/api/reports", tags=["reports"])

MANAGER = ("ADMIN", "ASSET_MANAGER")


@router.get("/utilization", response_model=list[UtilizationRow])
async def utilization(
    db: AsyncSession = Depends(get_db),
    _: Employee = Depends(require_roles(*MANAGER)),
):
    return await report_service.utilization(db)


@router.get("/maintenance-frequency", response_model=list[MaintenanceFrequencyRow])
async def maintenance_frequency(
    db: AsyncSession = Depends(get_db),
    _: Employee = Depends(require_roles(*MANAGER)),
):
    return await report_service.maintenance_frequency(db)


@router.get("/department-summary", response_model=list[DepartmentSummaryRow])
async def department_summary(
    db: AsyncSession = Depends(get_db),
    _: Employee = Depends(require_roles(*MANAGER)),
):
    return await report_service.department_summary(db)


@router.get("/booking-heatmap", response_model=BookingHeatmap)
async def booking_heatmap(
    db: AsyncSession = Depends(get_db),
    _: Employee = Depends(require_roles(*MANAGER)),
):
    return await report_service.booking_heatmap(db)


@router.get("/retirement-forecast", response_model=list[RetirementForecastRow])
async def retirement_forecast(
    older_than_years: int = 5,
    db: AsyncSession = Depends(get_db),
    _: Employee = Depends(require_roles(*MANAGER)),
):
    return await report_service.retirement_forecast(db, older_than_years)


@router.get("/export")
async def export(
    type: str = Query(..., description="utilization|maintenance-frequency|department-summary|retirement-forecast"),
    format: str = Query("xlsx", description="xlsx|pdf"),
    db: AsyncSession = Depends(get_db),
    _: Employee = Depends(require_roles(*MANAGER)),
):
    try:
        content, media_type, filename = await report_service.export_report(
            db, type, format
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    return StreamingResponse(
        io.BytesIO(content),
        media_type=media_type,
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: portal/src/lib/api/reports.ts
import { apiClient, API_BASE_URL } from "../api-client";
import { getAccessToken } from "../auth-store";

export interface UtilizationRow {
  asset_id: string;
  asset_tag: string;
  asset_name: string;
  allocation_count: number;
  total_allocation_days: number;
}

export interface MaintenanceFrequencyRow {
  asset_id: string;
  asset_tag: string;
  category_id: string | null;
  category_name: string | null;
  request_count: number;
  avg_resolution_hours: number | null;
}

export interface DepartmentSummaryRow {
  department_id: string | null;
  department_name: string | null;
  total_assets: number;
  allocated: number;
  available: number;
  under_maintenance: number;
}

export interface RetirementForecastRow {
  asset_id: string;
  asset_tag: string;
  asset_name: string;
  condition: string;
  acquisition_date: string | null;
  reason: string;
}

export interface BookingHeatmap {
  matrix: number[][];
}

export const reportsApi = {
  utilization: (signal?: AbortSignal) => apiClient.get<UtilizationRow[]>("/reports/utilization", { signal }),
  maintenanceFrequency: (signal?: AbortSignal) => apiClient.get<MaintenanceFrequencyRow[]>("/reports/maintenance-frequency", { signal }),
  departmentSummary: (signal?: AbortSignal) => apiClient.get<DepartmentSummaryRow[]>("/reports/department-summary", { signal }),
  retirementForecast: (signal?: AbortSignal) => apiClient.get<RetirementForecastRow[]>("/reports/retirement-forecast", { signal }),
  bookingHeatmap: (signal?: AbortSignal) => apiClient.get<BookingHeatmap>("/reports/booking-heatmap", { signal }),
  /** Downloads an export and triggers a browser save. */
  export: async (type: string, format: "xlsx" | "pdf" = "xlsx") => {
    const url = `${API_BASE_URL.replace(/\/$/, "")}/reports/export?type=${encodeURIComponent(type)}&format=${format}`;
    const res = await fetch(url, { headers: { Authorization: `Bearer ${getAccessToken() ?? ""}` } });
    if (!res.ok) throw new Error(`Export failed (${res.status})`);
    const blob = await res.blob();
    const disposition = res.headers.get("Content-Disposition") ?? "";
    const match = disposition.match(/filename="?([^"]+)"?/);
    const filename = match?.[1] ?? `${type}.${format}`;
    const objectUrl = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = objectUrl;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    a.remove();
    URL.revokeObjectURL(objectUrl);
  },
};
